src/twisted/test/test_tcp.py

```diff
@@ -120,9 +120,7 @@ class MyProtocolFactoryMixin:
 
     protocolConnectionMade = None
     protocolConnectionLost = None
-    protocol = (
-        None
-    )  # type: Optional[Union[Type[protocol.Protocol],Type[protocol.AbstractDatagramProtocol]]]  # noqa
```

Comment:
`protocol.AbstractDatagramProtocol` should be a `Protocol`. Or something.
Removing it from here and working around it with `type: ignore` (eg. see `test_tuntap.py`) to make it more obvious that it needs fixing.

Reply:
Yeah I had problems with AbstractDatagramProtocol when initially adding mypy support to the tree.

Reply:
Right.  When we add hints to `protocol.py` I think we need to figure out what to do with `AbstractDatagramProtocol`.  The current state of things seems incorrect to me, but I didn't want to spend to much time on that one just yet.

Reply:
`AbstractDatagramProtocol` has been around since about 2003: https://github.com/twisted/twisted/commit/869198ef461818414b794ae86cfc1e065ccea6fe

I'm not sure how easy it will be to "fix" a base class in a library that has been around so long, yet preserve backwards compatibility.  In my not be possible.

Reply:
Type signatures are not beholden to the compatibility policy, only runtime behavior.

Reply:
@glyph I think the issue is that `AbstractDatagramProtocol` is weird in that it does not subclass `Protocol`, and it seems to be the only example of that. But again, I didn't look into it.

Reply:
After having looked at it, it seems like AbstractDatagramProtocol is only expected here by the wacky twisted.pair tests, right?

Reply:
…correct

Reply:
Well, no… it's used in `twisted.pair` non-test code. But it's only in `twisted.pair`.

Reply:
I guess I'll need to run this for myself locally since I didn't see it in the mypy errors craig posted. My non-tool-assisted reading of the app code was that things take datagram protocols directly; do you have a pointer to where this is happening?

Reply:
I just did a search on the.sources:

```console
Searching 842 files for "AbstractDatagramProtocol" (case sensitive)

/Users/wsanchez/Dropbox/Developer/Twisted/twisted/src/twisted/internet/protocol.py:
  678  
  679  
  680: class AbstractDatagramProtocol:
  681      """
  682      Abstract protocol for datagram-oriented transports, e.g. IP, ICMP, ARP,
  ...
  754  
  755  @implementer(interfaces.ILoggingContext)
  756: class DatagramProtocol(AbstractDatagramProtocol):
  757      """
  758      Protocol for datagram-oriented transport, e.g. UDP.
  ...
  891      "FileWrapper",
  892      "ServerFactory",
  893:     "AbstractDatagramProtocol",
  894      "DatagramProtocol",
  895      "ConnectedDatagramProtocol",

/Users/wsanchez/Dropbox/Developer/Twisted/twisted/src/twisted/pair/ethernet.py:
   35  
   36  @implementer(IEthernetProtocol)
   37: class EthernetProtocol(protocol.AbstractDatagramProtocol):
   38      def __init__(self):
   39          self.etherProtos = {}

/Users/wsanchez/Dropbox/Developer/Twisted/twisted/src/twisted/pair/ip.py:
   44  
   45  @implementer(raw.IRawPacketProtocol)
   46: class IPProtocol(protocol.AbstractDatagramProtocol):
   47      def __init__(self):
   48          self.ipProtos = {}

/Users/wsanchez/Dropbox/Developer/Twisted/twisted/src/twisted/pair/rawudp.py:
   23  
   24  @implementer(raw.IRawDatagramProtocol)
   25: class RawUDPProtocol(protocol.AbstractDatagramProtocol):
   26      def __init__(self):
   27          self.udpProtos = {}

/Users/wsanchez/Dropbox/Developer/Twisted/twisted/src/twisted/pair/test/test_tuntap.py:
   38  from twisted.python.log import addObserver, removeObserver, textFromEventDict
   39  from twisted.internet.interfaces import IAddress, IReactorFDSet
   40: from twisted.internet.protocol import AbstractDatagramProtocol, Factory
   41  from twisted.internet.task import Clock
   42  from twisted.trial.unittest import SkipTest, SynchronousTestCase
   ..
 1305  
 1306  @implementer(IRawPacketProtocol)
 1307: class IPRecordingProtocol(AbstractDatagramProtocol):
 1308      """
 1309      A protocol which merely records the datagrams delivered to it.

8 matches across 5 files
```

Note also the subclass `DatagramProtocol`, which is used in even more places.

Reply:
Those seem to be usages of the class itself; what I meant by "expected here" (rather than "used") is that `Factory.protocol` (or <any-subclass-of-`Factory`>`.protocol`) is only expected to return `AbstractDatagramProtocol`s in that one case.  I do recognize that we have datagram protocols all over the place :)

Reply:
Going back up to the top of the thread here:

> `protocol.AbstractDatagramProtocol` should be a `Protocol`

Read from the perspective of "in the signature of `protocol.Factory`", then sure.  But from a higher-level perspective of "should the class exist": no, definitely not.  Its methods and lifecycle are totally incompatible.

The problem here is that `protocol.Factory` should never, ever return a `DatagramProtocol` (nor its unfortunately-named ancestor) because `protocol.Factory` is implicitly a factory for *stream-transport-oriented* Protocols, and `DatagramProtocol` is for processing unsequenced datagrams from a variety of sources; hence the fact that `datagramReceived` takes `addr` as well as `datagram`.

The reason it's `protocol.Factory` rather than `protocol.StreamFactory` on one side and `protocol.DatagramFactory` on the other is that datagram sockets do not get accepted to create connections; so there's no need for factory-level indirection or connection-level instances.  The socket just receives a datagram and the application gets told about it directly; that's it.  Hence, the various objects in `twisted.pair` and the reactor's own `listenUDP` just take an instance of a `DatagramProtocol` directly.

Reply:
OK, for that: https://twistedmatrix.com/trac/ticket/10013#ticket

Out of scope here.